algoliasearch/client.go

```diff
@@ -21,222 +19,187 @@ type Client struct {
+func (c *Client) AddKey(ACL []string, params map[string]interface{}) (res AddKeyRes, err error) {
+	req := make(map[string]interface{})
+	for k, v := range params {
+		req[k] = v
 	}
+	req["acl"] = ACL
 
-	return c.AddKeyWithParam(body)
-}
-
-// AddKeyWithParam creates a new API key using the specified
-// parameters.
-func (c *Client) AddKeyWithParam(params interface{}) (interface{}, error) {
-	return c.transport.request("POST", "/1/keys/", params, read)
-}
-
-// UpdateKey updates the API key named `key` using the `acl` operation
-// restrictions, the `indexes` the key is limited to, the `validity` during
-// which the key is to be valid in seconds, `maxQueriesPerIPPerHour` and
-// `maxHitsPerQuery` as rate limiters.
-func (c *Client) UpdateKey(key string, acl, indexes []string, validity int, maxQueriesPerIPPerHour int, maxHitsPerQuery int) (interface{}, error) {
-	body := map[string]interface{}{
-		"acl":                    acl,
-		"maxHitsPerQuery":        maxHitsPerQuery,
-		"maxQueriesPerIPPerHour": maxQueriesPerIPPerHour,
-		"validity":               validity,
-		"indexes":                indexes,
+	if err = checkKey(req); err != nil {
+		return res, err
 	}
 
-	return c.UpdateKeyWithParam(key, body)
+	err = c.request(&res, "POST", "/1/keys/", req, read)
+	return res, err
 }
 
 // UpdateKeyWithParam updates the API key named `key` with the supplied
 // parameters.
-func (c *Client) UpdateKeyWithParam(key string, params interface{}) (interface{}, error) {
-	return c.transport.request("PUT", "/1/keys/"+key, params, write)
+func (c *Client) UpdateKeyWithParam(key string, params map[string]interface{}) (res UpdateKeyRes, err error) {
+	if err = checkKey(params); err != nil {
+		return res, err
+	}
+
+	err = c.request(&res, "PUT", "/1/keys/"+key, params, write)
+	return res, err
 }
 
-// GetKey returns the characteristics of the API key named `key`.
-func (c *Client) GetKey(key string) (interface{}, error) {
-	return c.transport.request("GET", "/1/keys/"+key, nil, read)
+// GetKey returns the ACL and validity of the API key named `key`.
+func (c *Client) GetKey(key string) (res GetKeyRes, err error) {
+	err = c.request(&res, "GET", "/1/keys/"+key, nil, read)
+	return res, err
 }
 
 // DeleteKey deletes the API key named `key`.
-func (c *Client) DeleteKey(key string) (interface{}, error) {
-	return c.transport.request("DELETE", "/1/keys/"+key, nil, write)
+func (c *Client) DeleteKey(key string) (res DeleteKeyRes, err error) {
+	err = c.request(&res, "DELETE", "/1/keys/"+key, nil, write)
+	return res, err
 }
 
 // GetLogs retrieves the `length` latest logs, starting at `offset`. Logs can
 // be filtered by type via `logType` being either "query", "build" or "error".
-func (c *Client) GetLogs(offset, length int, logType string) (interface{}, error) {
-	body := map[string]interface{}{
-		"offset": offset,
-		"length": length,
-		"type":   logType,
+func (c *Client) GetLogs(params map[string]interface{}) (res GetLogsRes, err error) {
```

Comment:
After thoughts maybe I am against it. That would require taking pointers, so it would need the user to declare its values before the call. Don't know if it's that easier to use.